agent/lib_aws_sdk_php.c

```diff
@@ -0,0 +1,166 @@
+ * https://github.com/aws/aws-sdk-php
+ */
+#include "php_agent.h"
+#include "php_wrapper.h"
+#include "fw_hooks.h"
+#include "fw_support.h"
+#include "util_logging.h"
+#include "php_call.h"
+#include "lib_aws_sdk_php.h"
+
+#define PHP_PACKAGE_NAME "aws/aws-sdk-php"
+#define PHP_AWS_CLASS_PREFIX "Supportability/library/aws/aws-sdk-php/"
+
+/*
+ * In a normal course of events, the following line will always work
+ * zend_eval_string("Aws\\Sdk::VERSION;", &retval, "Get AWS Version")
+ * By the time we have detected the existence of the aws-sdk-php and with
+ * default composer profject settings, it callable even from
+ * nr_aws_sdk_php_enable which will automatically load the class if it isn't
+ * loaded yet and then evaluate the string. However, in the rare case that files
+ * are not loaded via autoloader and/or have non-default composer classload
+ * settings, if the class is not found, PHP 8.2+ will generate a fatal
+ * unrecoverable uncatchable error error whenever it cannot find a class. While
+ * calling this from nr_aws_sdk_php_enable would have been great and would allow
+ * the sdk version value to be set only once, to avoid the very unlikely but not
+ * impossible fatal error, this will be called from the
+ * "Aws\\ClientResolver::_apply_user_agent" wrapper which GUARANTEES that
+ * aws/sdk exists and is already loaded.
+ *
+ *
+ * Additionally given that aws-sdk-php is currently detected from the
+ * AwsClient.php file, this method will always be called when a client is
+ * created. Having separate functionality to extract from Aws/Sdk::__constructor
+ * is both not required and is redundant and causes additional overhead.
+ */
+extern void lib_aws_sdk_php_handle_version() {
+  char* version = NULL;
+  zval retval;
+  int result = FAILURE;
+
+  result = zend_eval_string("Aws\\Sdk::VERSION;", &retval, "Get AWS Version");
+
+  // Add php package to transaction
+  if (SUCCESS == result) {
+    if (nr_php_is_zval_non_empty_string(&retval)) {
+      version = Z_STRVAL(retval);
+    }
+    zval_dtor(&retval);
+  }
+
+  /* Use the Aws\Sdk::VERSION to determine the version */
+  if (NRINI(vulnerability_management_package_detection_enabled)) {
+    /* Add php package to transaction */
+    nr_txn_add_php_package(NRPRG(txn), PHP_PACKAGE_NAME, version);
+  }
+  nr_fw_support_add_package_supportability_metric(NRPRG(txn), PHP_PACKAGE_NAME,
+                                                  version);
+}
+
+extern void lib_aws_sdk_php_add_supportability_metric(const char* metric_name) {
+  char buf[512];
+  nrobj_t* names = NULL;
+
+  if (NULL == metric_name) {
+    return;
+  }
+  if (NULL == NRPRG(txn)) {
+    return;
+  }
+
+  /* Strip out the `\\` characters and normalize */
+  names = nr_strsplit(metric_name, "\\", 0 /* discard empty */);
+
+  buf[0] = '\0';
+  snprintf(buf, sizeof(buf), "%s%s", PHP_AWS_CLASS_PREFIX,
+           nro_get_array_string(names, 1, NULL));
+  for (int i = 2, n = nro_getsize(names); i <= n; i++) {
```

Comment:
```suggestion
  for (int i = 2, n = nro_getsize(names); i < n; i++) {
```
Maybe I am misunderstanding the loop, but I think it should be this?

Reply:
Ah I am misunderstanding.